File: backend/app/schemas/agent.py
from typing import Any

from pydantic import BaseModel, Field, field_validator


class AgentQueryRequest(BaseModel):
    session_id: int
    query: str = Field(min_length=1)
    max_retries: int | None = Field(default=None, ge=0, le=3)

    @field_validator("query")
    @classmethod
    def query_must_not_be_empty(cls, value: str) -> str:
        query = value.strip()
        if not query:
            raise ValueError("Query must not be empty")
        return query


class AgentTrace(BaseModel):
    intent: str | None = None
    planner_source: str | None = None
    response_source: str | None = None
    selected_tool: str | None = None
    validated_columns: list[str] = Field(default_factory=list)
    chart_url: str | None = None
    fallback_reason: str | None = None
    llm_enabled: bool | None = None
    llm_provider: str | None = None
    execution_mode: str | None = None
    code_generation_source: str | None = None
    safety_check: str | None = None
    runner: str | None = None
    timeout_seconds: int | None = None
    execution_status: str | None = None
    execution_time_ms: int | None = None
    reentry_used: bool | None = None
    retry_count: int | None = None
    max_retries: int | None = None
    first_attempt: dict[str, Any] | None = None
    repair_attempt: dict[str, Any] | None = None
    generated_code_preview: str | None = None
    error: str | None = None
    steps: list[str] = Field(default_factory=list)


class AgentQueryResponse(BaseModel):
    session_id: int
    dataset_id: int
    intent: str | None = None
    answer: str
    analysis_plan: list[str] = Field(default_factory=list)
    tool_used: str | None = None
    tool_result: dict[str, Any] = Field(default_factory=dict)
    chart_url: str | None = None
    follow_up_questions: list[str] = Field(default_factory=list)
    agent_trace: AgentTrace | None = None
    execution_mode: str | None = None
    code_execution_result: dict[str, Any] | None = None
    generated_code_preview: str | None = None
    metadata: dict[str, Any]

File: frontend/src/types/agent.ts
export interface AgentTrace {
  intent?: string;
  planner_source?: string;
  response_source?: string;
  selected_tool?: string;
  validated_columns?: string[];
  chart_url?: string | null;
  fallback_reason?: string | null;
  llm_enabled?: boolean;
  llm_provider?: string | null;
  execution_mode?: string | null;
  code_generation_source?: string | null;
  safety_check?: string | null;
  runner?: string | null;
  timeout_seconds?: number | null;
  execution_status?: string | null;
  execution_time_ms?: number | null;
  reentry_used?: boolean | null;
  retry_count?: number | null;
  max_retries?: number | null;
  first_attempt?: Record<string, unknown> | null;
  repair_attempt?: Record<string, unknown> | null;
  generated_code_preview?: string | null;
  error?: string | null;
  steps?: string[];
}

export type AgentQueryResponse = {
  session_id: number;
  dataset_id: number;
  intent?: string;
  answer: string;
  analysis_plan?: string[];
  tool_used?: string | null;
  tool_result?: unknown;
  chart_url?: string | null;
  follow_up_questions?: string[];
  agent_trace?: AgentTrace;
  execution_mode?: string | null;
  code_execution_result?: Record<string, unknown> | null;
  generated_code_preview?: string | null;
  metadata?: Record<string, unknown>;
};
